### PR TITLE
Calculate average function

### PR DESCRIPTION
**Pull request by Yevheniia-Rudenko**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Yevheniia-Rudenko** [comment]
Hi everyone! This is my first attempt of the pull request. Could someone please review my code?

**Comment by Awaab98** [comment]
I will take on the review process.

**Comment by Awaab98** [comment]
Lets start with the CI checks, then I will comment on the code. We have three CI checks on push, and three simillar ones on the PR, which did not pass. This means that we only have to pass three checks. I have some suggestions for each one.

**Comment by Awaab98** [comment]
As for the first faild check: **md_formating**, you can solve this by going to **collaboration/guide/0_repository_setup.md** file in your branch repository, **line 17**, and break the line into two lines since its length is 132, which is more than the maximum expected line length of 80. This way the **md_formating** check will pass.

**Comment by Awaab98** [comment]
Now, for the both **py_formatting** faild checks, install **ruff** locally. Then use the command: **ruff check file.py** to ensure that both the function and the tests are compliant with ruff's linting standards. Once the checks pass locally, commit and push the changes, and the CI check will pass.

**Comment by Yevheniia-Rudenko** [comment]
@Awaab98 Thank you so much for your review and your suggestions! I will work on addressing the CI checks!! I’ll let you know as soon as I do any changes.

**Comment by Allukmon** [comment]
This is really helpful, I am able to get some meaningful insight that will help me.

Thanks a lot @Yevheniia-Rudenko @Awaab98

**Comment by Awaab98** [comment]
@Yevheniia-Rudenko I have updated your pull request with the latest changes from main. Now, it is **8/10** on the CI checks you only need to work on the the **py_formating**  CI check as I explanied in the previous comments.

**Comment by Yevheniia-Rudenko** [comment]
I tried to update ruff to the latest version to use the --fix option for automatically fixing code formatting. The installed version (0.8.4) doesn't support it. However, all ruff checks on my local machine worked fine, and the code is correctly formatted locally.

When I tried to install the latest version from GitHub, I got an error because Rust and Cargo are not installed. I will install Rust and Cargo, then try installing ruff again and check for the --fix option.

**Comment by Awaab98** [comment]
@Yevheniia-Rudenko. You are right. I have updated your pull request with the latest changes from main after the latest changes introduced by Evan to the settings.json, commit: 8e42c8768ebeadede5391f31dc632e3693f995f2 , **use ruff as formatter**. Still, the two checks relating to **py_formatting** faild. This issue is related to the formatting of the both `.py` files, hence, I will check your files and suggest the solving changes accordingly.

**Comment by Awaab98** [comment]
@Yevheniia-Rudenko. **Good news!** I was able to figure out a solution to the **py_formating** CI CHECK. You can find the suggested changes as comments to the respective lines on both the function and the test files. Once you integrate the suggested changes, the aforementioned check will succeed.

**Comment by Yevheniia-Rudenko** [comment]
@Awaab98, the CI checks issue is finally resolved—it was easier than I expected! I updated pip and ruff and ran the ruff format command. Thank you so much for your help and guidance!

**Comment by Awaab98** [comment]
@Yevheniia-Rudenko Thats great. I see that you have made some changes to the **settings.json** file. I suggest that you consider the changes to the empty lines that I have made because they allow you to pass the **py_formatting** check without the need to change the setting.json file. You can find my detailed suggestions on the **files changed** tab.

**Comment by Yevheniia-Rudenko** [comment]
@Awaab98, all done—I reverted to the initial VS Code settings, and the CI checks are still passing. Thanks for your suggestions and support! 🙌

**Review by Awaab98 — CHANGES_REQUESTED** [review]
@Yevheniia-Rudenko I have suggested some changes as you can see under the files changed section.

File: solutions/calculate_average.py
[the remark this answers is not in the document]
Comment (Yevheniia-Rudenko) [reply solutions/calculate_average.py:1]: Thank you for pointing this out! I have removed all unnecessary indentations on empty lines to ensure the file adheres to proper formatting standards and passes the Ruff check.

File: solutions/calculate_average.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+Module contents:
+    - calculate_average: calculates the average of two numbers.
+
+Created on 26.12.2024
+@author: Yevheniia Rudenko
+"""
+
+
+def calculate_average(num1: float, num2: float) -> float:
```
[the remark this answers is not in the document]
Comment (Yevheniia-Rudenko) [reply solutions/calculate_average.py:14]: Agreed. I'll update num1 to the_first_number and num2 to the_second_number throughout the code and documentation in the next commit.

File: solutions/calculate_average.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    Parameters:
+        num1: float, the first number
+        num2: float, the second number
+
+    Returns -> float: the average of num1 and num2
+
+    Raises:
+        AssertionError: if any argument is not a number
+
```
[the remark this answers is not in the document]
Comment (Yevheniia-Rudenko) [reply solutions/calculate_average.py:25]: Adding an Examples section is a good idea. It will be included in the function docstring in the next commit.

File: solutions/calculate_average.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    """Calculate the average of two numbers.
+
+    Parameters:
+        num1: float, the first number
+        num2: float, the second number
+
+    Returns -> float: the average of num1 and num2
+
+    Raises:
+        AssertionError: if any argument is not a number
```
[the remark this answers is not in the document]
Comment (Yevheniia-Rudenko) [reply solutions/calculate_average.py:24]: TypeError is more appropriate for type validation, you are right. This change will be implemented in the next commit.

File: solutions/tests/test_calculate_average.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Tests for calculate_average function.
+
```
[the remark this answers is not in the document]
Comment (Yevheniia-Rudenko) [reply solutions/tests/test_calculate_average.py:5]: I will add the summary for the test categories of the test module in the next commit.

**Comment by Yevheniia-Rudenko** [comment]
Hello, @Awaab98 Thank you so much for your comments, it was really helpful! I pushed some changes based on your review, could you please take a look? 55581a456ce2f8fb6ba43e2e27103a1a3f0a3aaa

**Comment by Awaab98** [comment]
fj

**Review by Awaab98 — APPROVED** [review]
Yes. That is great. I have checked the changes that you implemented. Everythings stands in order, and I am going to merge the pull request.

**Comment by Yevheniia-Rudenko** [comment]
Thank you @Awaab98 !
